docs/mergeop.md

```diff
@@ -0,0 +1,287 @@
+# Requirements
+1. "Fast" merges that don't duplicate data on disk when using the right underlying snapshotter and have an execution time less than or equal to the previous llb.Copy implementation in vast majority of cases.
+   * For copies specifically, `llb.Copy(src, dest)` must have the same complexity as `Merge(dest, llb.Copy(src, scratch))`, as described more below
+1. Fallback merges that work with any snapshotter, also with execution time less than or equal to previous llb.Copy implementation
+1. Merges are lazy in that they are not created as snapshots on disk until strictly needed.
+1. When exporting an image created by merging in a separate llb.State, identical layers are not re-exported, they are re-used from their merged source.
+   * If pushing to a remote registry that already has a layer input to a merge, no new layer is pushed
+   * If a merge input doesn't exist locally (it's lazy) but does exist in the remote being pushed to, it stays lazy
+1. Invalidation of one merge input does not result in other independent merge inputs being invalidated
+   * Tonis's specific requirements around this are included as Examples below.
+1. Support for optimizing Dockerfile `COPY` statements
+1. Works with DiffOp
+
+# Design
+## `MergeOp`
+1. MergeOp will be a new op that accepts multiple inputs and attempts to merge them in such a way as to minimize execution time and the amount of data duplicated on disk.
+   * For the first iteration, merging will be implemented by creating hardlinks (as discussed more below). If hardlinks are not supported (perhaps due to a filesystem underlying the snapshotter not supporting them), then it will fallback to a less efficient copy.
```

Comment:
All the main ones (overlay, native, stargz) do yeah. I don't remember enough details of how some of the other ones like devmapper work and haven't ever looked into others like btrfs, so I added this caveat since AFAIK it's now possible to use any arbitrary snapshotter because buildkit supports remote snapshotters.